pkg/volume/csi/csi_plugin.go

```diff
@@ -246,13 +247,104 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
 		}
 	}
 
-	// Initializing csiDrivers map and label management channels
+	// Initializing csiDrivers map, label management channels, and migratedDrivers map
 	csiDrivers = csiDriversStore{driversMap: map[string]csiDriver{}}
-	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host)
 
-	// TODO(#70514) Init CSINodeInfo object if the CRD exists and create Driver
-	// objects for migrated drivers.
+	var migratedDrivers = map[string](func() bool){
+		csilibplugins.GCEPDDriverName: func() bool {
+			return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) && utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE)
+		},
+		// TODO(leakingtpan): Add AWS migration feature gates and place them here
+	}
+
+	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host, migratedDrivers)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		// This function prevents Kubelet from posting Ready status until CSINodeInfo
+		// is both installed and initialized
+		err := initializeCSINodeInfo(host)
+		if err != nil {
+			return fmt.Errorf("failed to initialize CSINodeInfo: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func initializeCSINodeInfo(host volume.VolumeHost) error {
+	kvh, ok := host.(volume.KubeletVolumeHost)
+	if !ok {
+		klog.V(4).Infof("Skipping CSINodeInfo initialization, not running on Kubelet")
+		return nil
+	}
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil {
+		// If we're NOT on the Kubelet OR KubeClient DOESNT exist
+		return fmt.Errorf("failed to get KubeClient")
+	}
+
+	kvh.SetKubeletError(fmt.Errorf("CSINodeInfo is not yet available"))
+
+	go func() {
+		defer utilruntime.HandleCrash()
+
+		// Backoff parameters tuned to retry over 140 seconds. Will fail and restart the Kubelet
```

Comment:
cc @smarterclayton @kubernetes/sig-cluster-lifecycle for possible implications on self-hosted cases... how long should the api-contacting portion of a kubelet wait before giving up and triggering kubelet exit?

Reply:
/assign @timothysc 

from my perspective it seems wrong for a CSI plugin or any plugable solution to manage a kublet restart in the first place. the kubelet is a system service.
kubelet restarts should be managed by higher level tools that the user invokes close to the OS.

i'm finding it difficult understanding why this is required from `csi_plugin.go`?

> how long should the api-contacting portion of a kubelet wait before giving up and triggering kubelet exit?

kubeadm is doing 40seconds for the healthz endpoint check:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/wait.go#L164

Reply:
`csi_plugin.go` is not the pluggable solution itself. It is the system piece that will connect to the various CSI Drivers that are eventually plugged in to Kubernetes.

This init step will be *required* to "boot the volume subsystem" for future versions of Kubernetes.

Reply:
> kubeadm is doing 40seconds for the healthz endpoint check:

correction. it's 155 seconds, but there was an initial timeout of 40 seconds.

> csi_plugin.go is not the pluggable solution itself. It is the system piece that will connect to the various CSI Drivers that are eventually plugged in to Kubernetes.

i guess i was misled by the file name, as it implies a plugin.

Reply:
Currently, kubelet doesn't fail if it has problems reaching the apisver. Maybe we should keep that behavior unless there is clear benefit of restarting kubelet.